packages/core/src/common/message-rpc/rpc-message-encoder.ts

```diff
@@ -476,22 +526,23 @@ export class RpcMessageEncoder {
         this.writeTypedValue(buf, err);
     }
 
-    writeTypedValue(buf: WriteBuffer, value: any): void {
+    writeTypedValue(buf: WriteBuffer, value: any, visitedObjects = new WeakSet()): void {
         for (let i: number = this.encoders.length - 1; i >= 0; i--) {
             if (this.encoders[i][1].is(value)) {
                 buf.writeUint8(this.encoders[i][0]);
-                this.encoders[i][1].write(buf, value, (innerBuffer, innerValue) => {
-                    this.writeTypedValue(innerBuffer, innerValue);
+                this.encoders[i][1].write(buf, value, visitedObjects, (innerBuffer, innerValue, _visitedObjects) => {
+                    this.writeTypedValue(innerBuffer, innerValue, _visitedObjects);
                 });
                 return;
             }
         }
+        throw new EncodingError(`No suitable value encoder found for ${value}`);
     }
 
-    writeArray(buf: WriteBuffer, value: any[]): void {
+    writeArray(buf: WriteBuffer, value: any[], visitedObjects = new WeakSet()): void {
```

Comment:
I would feel more comfortable not using a default argument here: without it, we would catch cases where the "visitedObjects" are not forwarded.